core/src/test/java/org/elasticsearch/search/aggregations/BaseAggregationTestCase.java

```diff
@@ -0,0 +1,259 @@
+        XContentParser parser = XContentFactory.xContent(contentString).createParser(contentString);
+        assertSame(XContentParser.Token.START_OBJECT, parser.nextToken());
+        assertSame(XContentParser.Token.FIELD_NAME, parser.nextToken());
+        assertEquals(testAgg.name, parser.currentName());
+        assertSame(XContentParser.Token.START_OBJECT, parser.nextToken());
+        assertSame(XContentParser.Token.FIELD_NAME, parser.nextToken());
+        assertEquals(testAgg.type, parser.currentName());
+        assertSame(XContentParser.Token.START_OBJECT, parser.nextToken());
+        AggregatorFactory newAgg = aggParsers.parser(testAgg.getWriteableName()).parse(testAgg.name, parser, SearchContext.current());
+        assertSame(XContentParser.Token.END_OBJECT, parser.currentToken());
+        assertSame(XContentParser.Token.END_OBJECT, parser.nextToken());
+        assertSame(XContentParser.Token.END_OBJECT, parser.nextToken());
+        assertNull(parser.nextToken());
+        assertNotNull(newAgg);
+        assertNotSame(newAgg, testAgg);
+        assertEquals(testAgg, newAgg);
+        assertEquals(testAgg.hashCode(), newAgg.hashCode());
+    }
+
+    /**
+     * Test serialization and deserialization of the test AggregatorFactory.
+     */
+    @Test
+    public void testSerialization() throws IOException {
+        AF testAgg = createTestAggregatorFactory();
+        try (BytesStreamOutput output = new BytesStreamOutput()) {
+            testAgg.writeTo(output);
+            try (StreamInput in = new NamedWriteableAwareStreamInput(StreamInput.wrap(output.bytes()), namedWriteableRegistry)) {
+                AggregatorFactory prototype = aggParsers.parser(testAgg.getWriteableName()).getFactoryPrototype();
+                AggregatorFactory deserializedQuery = prototype.readFrom(in);
+                assertEquals(deserializedQuery, testAgg);
+                assertEquals(deserializedQuery.hashCode(), testAgg.hashCode());
+                assertNotSame(deserializedQuery, testAgg);
+            }
+        }
+    }
+
+    protected String[] getRandomTypes() {
+        String[] types;
+        if (currentTypes.length > 0 && randomBoolean()) {
+            int numberOfQueryTypes = randomIntBetween(1, currentTypes.length);
+            types = new String[numberOfQueryTypes];
+            for (int i = 0; i < numberOfQueryTypes; i++) {
+                types[i] = randomFrom(currentTypes);
+            }
+        } else {
+            if (randomBoolean()) {
+                types = new String[] { MetaData.ALL };
+            } else {
+                types = new String[0];
+            }
+        }
+        return types;
+    }
+
+    public String randomNumericField() {
+        int randomInt = randomInt(3);
+        switch (randomInt) {
+        case 0:
+            return DATE_FIELD_NAME;
+        case 1:
+            return DOUBLE_FIELD_NAME;
+        case 2:
+        default:
+            return INT_FIELD_NAME;
+        }
```

Comment:
One more thing, maybe you want to have a testEqualsAndHashcode like we recently introduced in the query-refactoring branch, just to make sure that equals and hashcode are properly implemented.